rfd/0180-sso-as-mfa-method.md

```diff
@@ -0,0 +1,528 @@
+
+When device trust is required, newly auto-provisioned SSO users are required
+to add their first MFA device from a trusted device. When combined with MFA
+security features, such as MFA for Admin actions and per-session MFA, the blast
+radius of an IdP compromise is largely contained. The attacker would be
+prevented from accessing any critical infrastructure or making any changes
+to the cluster's security configuration.
+
+Allowing SSO as an MFA method would bypass the device trust check, opening the
+cluster back up to attacks in the case of an IdP compromise. To maintain this
+invariant, device trust must be enforced within the SSO MFA check.
+
+### Implementation
+
+#### Privilege Tokens
+
+Privileged tokens are used as transient MFA verification for some operations in
+Teleport today, like account resets. These tokens can be generated by passing a
+completed MFA challenge to `rpc CreatePrivilegeToken`.
+
+Currently, privilege tokens can only be used in select endpoints as a replacement
+for MFA. Privilege Tokens will be supported as a normal MFA response to support
+other use cases:
+
+```proto
+message MFAAuthenticateResponse {
+  oneof Response {
+    TOTPResponse TOTP = 2;
+    webauthn.CredentialAssertionResponse Webauthn = 3;
+    PrivilegeTokenResponse Token = 4;
+  }
+}
+```
+
+Note: Privilege tokens used in this way must be consumed upon verification to
+ensure that exchanging a webauthn or totp token for a privilege token through
+`rpc CreatePrivilegeToken` has no significant security implications.
+
+#### SSO MFA flow
+
+We will introduce a new SSO login flow that allows clients to get a new privilege
+token instead of the standard login credentials. Authenticated clients can create
+an sso auth request with `rpc CreateSAMLAuthRequest`, `rpc CreateOIDCAuthRequest`
+or `rpc CreateGithubAuthRequest` and set `CreatePrivilegeToken=true` to use this
+new flow. Once the client authenticates through the IdP, the client will be given
+a privilege token generated for the user.
+
+Note: As is the case with normal SSO login, the login response is encrypted
+using a secret key owned by the client so that the token can not be intercepted
+in a man in the middle attack.
+
+#### Forced Re-authentication
+
+By default, Teleport will require the user to re-login through their IdP to
+pass an MFA check. This will prevent long lasting login sessions from being
+used as MFA verification, which could easily be exploited by an attacker with
+remote access to the user's machine.
+
+For SAML, this can be achieved by setting `ForceAuthn=true` in the authn
+request. For OIDC and Github, setting `max_age=0` in the server redirect URL's
+query parameter will force re-auth.
+
+In cases where re-auth is not desired behaviour, such as when the IdP is
+configured to prompt for MFA for existing sessions, users can set `force_authn`
+or `max_age` in the SAML or OIDC connector respectively to override the default
+behavior.
```

Comment:
If the values are present in the connector then we'll not set them?

I'm a bit concerned that a misconfiguration here could lead to do-nothing MFA flows.

Reply:
Poor wording by me, they must set `force_authn=false` or `max_age=8h` to disable re-auth.